README
This is the special repository with translations for the Psi+ and Psi projects:
  https://psi-plus.com/
  https://psi-im.org/

Homepage: https://github.com/psi-plus/psi-plus-l10n

Translations are managed via Transifex web service:
  https://www.transifex.com/tehnick/psi-plus/

If you want to begin new translation or update an existing one you should simply
register in Transifex and use its standard mechanisms for creating of new
translation or for joining translators team.

You are free to choose the way in which you want to update the translation:
* you may edit translation file via Transifex web interface from everywhere
* you may download translation file, edit it in any editor which you like
  (we recommend Qt Linguist for this) and upload updated file back

Transifex has quite good documentation:
  https://help.transifex.com/en/

If you want use only web UI you do not need to read anything because web UI is
very simple and you may just begin to work.

But if you want to manage translations manually you may use special command-line
program transifex-client. It has simple syntax similar to git:
  https://developers.transifex.com/docs/cli

This repo was created by Boris Pek <tehnick-8@yandex.ru>. Please contact me if
you have any troubles with this repo.

If you want to fix one or two words in some translation, you may just write
about them here: https://github.com/psi-plus/psi-plus-l10n/issues
But for more amount of changes please join to our translators team.

If you find typos in the program sources you may report bug here:
  https://github.com/psi-im/psi/issues

Useful links:
  https://github.com/psi-im/psi-l10n
  https://sourceforge.net/projects/psiplus/files/Translations/


ChangeLog
2023-02-26:
* Besnik Bleta is now in Albanian translation team.

2023-02-23:
* Update script: migrate to new transifex client

2023-02-03:
* Added template for Albanian translation.

2020-12-20:
* Anh Nguyen joined to Vietnamese translation team.

2017-08-06:
* Update script:
  add new command line option "desktop_up" (update main desktop file)

2017-06-28:
* Myscelus is now in Bulgarian translation team.
* Morgan' GAILLARD is now in French translation team.
* Christian Weiske is now in German translation team.

2017-06-16:
* Update script: small cosmetic changes.

2016-05-09:
* Aldo Gate is now in Italian translation team.

2016-04-06:
* Update script:
  - update URL for psi-plus-snapshots git repo
  - rewrite from bash to sh

2016-03-31:
* Sandor Matyi is now in Hungarian translation team.

2015-09-01:
* Update script:
  - add new command line option "tr_fu" (fast update)
  - add the most important comments

2015-07-02:
* Update script: more verbose output for "tarball" option.

2014-11-12:
* Aurélien TAMISIER is now in French translation team.

2014-04-19:
* Стоян Димитров is now in Bulgarian translation team.

2014-03-17:
* Update script: get strings from all extra patches.

2014-02-04:
* Update script: delete link to Psi+ project at code.google.com

2014-01-18:
* Script was updated: add link to Psi+ project at sourceforge.net

2014-01-11:
* Florian Fieber is now in German translation team.

2013-11-24:
* Apollo is now in Chinese Traditional translation team.

2013-11-22:
* Apollo is now in Chinese Simplified translation team.

2013-10-14:
* Script was updated: small fix due to changes in psi-plus-snapshots location.

2013-07-17:
* Script was updated: do not delete psi_en.qm
  (English translation is required for some template strings)

2013-06-01:
* Script was updated: more verbose for tag option.
* comicfans44 is now in Chinese Simplified translation team.

2013-05-28:
* Script was updated: hostname transifex.net was changed to transifex.com.

2013-05-27:
* Ernesto Avilés Vázquez is now in Spanish translation team.
* LukynZ is now in Czech translation team.

2013-05-14:
* Jakub Liput is now in Polish translation team.

2013-04-04:
* oMid Hedayati is now in Persian translation team.

2013-02-07:
* Alexandro Casanova is now in Brazilian Portuguese translation team.

2013-02-01:
* Baurzhan Muftakhidinov is now in Kazakh translation team.

2013-01-14:
* Script was updated:
  - updated command line option 'tarball'; code.google.com will be used for
    sharing precompiled localization files because github.com closed this
    feature
  - changed name format for tarball with precompiled localization files
    (see https://psi-plus.com/wiki/en:packagement_rules)

2012-11-16:
* Script was updated: small fix in tr_sync option.
* Isratine Citizen is now in Hebrew translation team.

2012-11-12:
* Script was updated: fixed a typo.

2012-10-17:
* Script was updated:
  - new format of command line option tr_push
  - fixed offline work

2012-10-14:
* Added non-translated localization file for Hebrew translation.

2012-10-12:
* Script was updated:
  fixed determination of CUR_DIR variable when absolute path is used.

2012-10-11:
* Script was updated:
  - added new command line option tr_sync
  - test Internet connection

2012-10-05:
* Added non-translated localization file for Arabic translation.

2012-09-26:
* Script was updated:
  - Git repo was moved to new location (https://github.com/psi-plus)
  - psi-plus-i18n was renamed to psi-plus-l10n

2012-08-24:
* Script was updated: added new command line option tr_push.

2012-06-27:
* Jiri Slaby is now in Czech translation team.

2012-05-24:
* Vitaly Tonkacheyev is now in Ukrainian translation team.

2012-05-10:
* edasmax is now in Polish translation team.

2012-04-06:
* Michal Hrušecký is now in Czech translation team.

2012-04-02:
* Erik-B. Ernst is now in German translation team.
* Benjamin Weber is now in German translation team.
* Márk Lutring is now in Hungarian translation team.
* Guillermo Castellá is now in Spanish translation team.
* Pavał Šałamicki is now in Belarusian translation team.
* Miroslav Petrovic is now in Serbian (Latin) translation team.

2012-03-24:
* Repository was created.
* Added old translations from Psi IM which were updated using actual sources
  of Psi+.
* Russian translation of Psi+ was taken from special repository
  (https://github.com/ivan101/psi-plus-ru)


update-repo.sh
#!/bin/sh

# Author:  Boris Pek <tehnick-8@mail.ru>
# License: GPLv2 or later
# Created: 2012-03-24
# Updated: 2016-05-25
# Version: N/A

export CUR_DIR="$(dirname $(realpath -s ${0}))"
export MAIN_DIR="${CUR_DIR}/.."
export PSIPLUS_DIR="${MAIN_DIR}/psi-plus-snapshots"

cd "${CUR_DIR}" || exit 1

case "${1}" in
"up")
    # Pulling changes from GitHub repo.

    git pull --all || exit 1

;;
"cm")
    # Creating correct git commit.

    git commit -a -m 'Translations were updated from Transifex.' || exit 1

;;
"tag")
    # Creating correct git tag.

    cd "${PSIPLUS_DIR}" || exit 1
    CUR_TAG="$(git tag -l  | sort -r -V | head -n1)"

    cd "${CUR_DIR}" || exit 1
    echo "git tag \"${CUR_TAG}\""
    git tag "${CUR_TAG}"

    echo ;
    echo "Last tags:"
    git tag | sort -V | tail -n 10

;;
"push")
    # Pushing changes into GitHub repo.

    git push || exit 1
    git push --tags || exit 1

;;
"make")
    # Making precompiled localization files.

    rm translations.pro

    echo "TRANSLATIONS = \\" >> translations.pro
    echo translations/*.ts >> translations.pro

    lrelease ./translations.pro

    mkdir -p out
    mv translations/*.qm out/ || exit 1

;;
"install")
    # Installing precompiled localization files into default directory.

    if [ ${USER} != "root" ]; then
        echo "You are not a root now!"
        exit 1
    fi

    mkdir -p /usr/share/psi-plus/translations/
    cp out/*.qm /usr/share/psi-plus/translations/ || exit 1

;;
"tarball")
    # Generating tarball with precompiled localization files.

    CUR_TAG="$(git tag -l  | sort -r -V | head -n1)"

    rm -rf psi-plus-translations-*
    mkdir psi-plus-translations-${CUR_TAG} || exit 1
    cp out/*.qm psi-plus-translations-${CUR_TAG} || exit 1

    tar -cJf psi-plus-translations-${CUR_TAG}.tar.xz psi-plus-translations-${CUR_TAG} || exit 1
    echo "Tarball with precompiled translation files is ready for upload:"
    [ ! -z "$(which realpath)" ] && echo "$(realpath ${CUR_DIR}/psi-plus-translations-${CUR_TAG}.tar.xz)"
    echo "https://sourceforge.net/projects/psiplus/files/Translations/"

;;
"tr")
    # Pulling changes from Transifex.

    # Test Internet connection:
    host transifex.com > /dev/null || exit 1

    git status || exit 1

    LANG_DIR="${MAIN_DIR}/psi-plus-l10n_transifex"
    cd "${LANG_DIR}" || exit 1
    tx pull -a -s || exit 1

    cd "${LANG_DIR}/translations/psi-plus.full/" || exit 1
    cp *.ts "${CUR_DIR}/translations/"

    cd "${LANG_DIR}/translations/psi-plus.desktop-file/" || exit 1
    cp *.desktop "${CUR_DIR}/desktop-file/"

    cd "${CUR_DIR}" || exit 1
    git status || exit 1

;;
"tr_up")
    # Full update of localization files.

    git status || exit 1

    if [ -d "${PSIPLUS_DIR}" ]; then
        echo "Updating ${PSIPLUS_DIR}"
        cd "${PSIPLUS_DIR}"
        git pull --all || exit 1
        echo;
    else
        echo "Creating ${PSIPLUS_DIR}"
        cd "${MAIN_DIR}"
        git clone https://github.com/psi-plus/psi-plus-snapshots.git || exit 1
        echo;
    fi

    # beginning of magical hack
    cd "${CUR_DIR}"
    rm -fr tmp
    mkdir tmp
    cd tmp/

    cp "${PSIPLUS_DIR}/patches"/*/*.diff ./ || exit 1
    cp "${PSIPLUS_DIR}/patches"/*/*.patch ./ || exit 1
    PATCHES=$(ls *.diff *.patch)
    FILES="$(grep '^--- a/' ${PATCHES} | sed -e 's|^.*:--- a/\(.*\)$|\1|' | sort -u)
           $(grep '^--- psi.orig/' ${PATCHES} | sed -e 's|^.*:--- psi.orig/\(.*\)$|\1|' | sort -u)"
    DIRS="$(dirname ${FILES} | sort -u | grep -v "^\.$")"
    for DIR in ${DIRS} ; do
        mkdir -p ${DIR}
    done
    for FILE in ${FILES} ; do
        cp -f ${PSIPLUS_DIR}/${FILE} "${FILE}" 2>/dev/null
    done
    for PATCH in ${PATCHES} ; do
        patch -f -p1 < "${PATCH}" > applied_patches.log
    done
    rm ${PATCHES}

    cd "${PSIPLUS_DIR}/src"
    python ../admin/update_options_ts.py ../options/default.xml > \
        "${CUR_DIR}/tmp/option_translations.cpp"
    # ending of magical hack

    cd "${CUR_DIR}"
    rm translations.pro

    echo "HEADERS = \\" >> translations.pro
    find "${PSIPLUS_DIR}/iris" "${PSIPLUS_DIR}/src" "${CUR_DIR}/tmp" -type f -name "*.h" | \
        while read var; do echo "  ${var} \\" >> translations.pro; done

    echo "SOURCES = \\" >> translations.pro
    find "${PSIPLUS_DIR}/iris" "${PSIPLUS_DIR}/src" "${CUR_DIR}/tmp" -type f -name "*.cpp" | \
        while read var; do echo "  ${var} \\" >> translations.pro; done
    echo "  ${CUR_DIR}/tmp/*.cpp" >> translations.pro

    echo "FORMS = \\" >> translations.pro
    find "${PSIPLUS_DIR}/iris" "${PSIPLUS_DIR}/src" "${CUR_DIR}/tmp" -type f -name "*.ui" | \
        while read var; do echo "  ${var} \\" >> translations.pro; done
    echo "  ${CUR_DIR}/tmp/*.ui" >> translations.pro

    echo "TRANSLATIONS = \\" >> translations.pro
    echo translations/*.ts >> translations.pro

    lupdate -verbose ./translations.pro

    cp "${PSIPLUS_DIR}"/*.desktop "${CUR_DIR}/desktop-file/"

    git status || exit 1

;;
"tr_fu")
    # Fast update of localization files.

    git status || exit 1

    lupdate -verbose ./translations.pro

    cp "${PSIPLUS_DIR}"/*.desktop "${CUR_DIR}/desktop-file/"

    git status || exit 1

;;
"tr_cl")
    # Cleaning update of localization files.

    git status || exit 1

    lupdate -verbose -no-obsolete ./translations.pro

    cp "${PSIPLUS_DIR}"/*.desktop "${CUR_DIR}/desktop-file/"

    git status || exit 1

;;
"tr_push")
    # Pushing changes to Transifex.

    LANG_DIR="${MAIN_DIR}/psi-plus-l10n_transifex/"
    cd "${LANG_DIR}/translations" || exit 1

    cp "${CUR_DIR}"/translations/*.ts psi-plus.full/
    cp "${CUR_DIR}"/desktop-file/*.ts psi-plus.desktop-file/

    cd "${LANG_DIR}"
    if [ -z "${2}" ]; then
        echo "<arg> is not specified!"
        exit 1
    elif [ "${2}" = "src" ] ; then
        tx push -s || exit 1
    elif [ "${2}" = "all" ] ; then
        tx push -s -t || exit 1
    else
        tx push -t -l ${2} || exit 1
    fi

;;
"tr_co")
    # Cloning Transifex repo.

    if [ -d "${MAIN_DIR}/psi-plus-l10n_transifex" ]; then
        echo "${MAIN_DIR}/psi-plus-l10n_transifex"
        echo "directory already exists!"
    else
        echo "Creating ${MAIN_DIR}/psi-plus-l10n_transifex"
        mkdir -p "${MAIN_DIR}/psi-plus-l10n_transifex/.tx"
        cp "transifex.config" "${MAIN_DIR}/psi-plus-l10n_transifex/.tx/config"
        cd "${MAIN_DIR}/psi-plus-l10n_transifex" || exit 1
        tx pull -a -s || exit 1
    fi

;;
"tr_sync")
    # Syncing Transifex and Guthub repos.

    "${0}" up > /dev/null || exit 1
    "${0}" tr > /dev/null || exit 1

    if [ "$(git status | grep 'translations/' | wc -l)" -gt 0 ]; then
        "${0}" cm || exit 1
        "${0}" push || exit 1
    fi
    echo ;
;;
*)
    # Help.

    echo "Usage:"
    echo "  up cm tag push make install tarball"
    echo "  tr tr_up tr_fu tr_cl tr_co tr_sync"
    echo "  tr_push <arg> (arg: src, all or language code)"
    echo ;
    echo "Examples:"
    echo "  ./update-repo.sh tr_push src"
    echo "  ./update-repo.sh tr_push all"
    echo "  ./update-repo.sh tr_push ru"

;;
esac

exit 0
